File: schemas/media-streaming/mongodb/albums.js
// albums: Music albums, singles, EPs, and compilations with metadata and popularity.
// See README.md for full design rationale.

const mongoose = require("mongoose");

const albumsSchema = new mongoose.Schema(
  {
    title: { type: String, required: true },
    slug: { type: String, unique: true, required: true },
    artist_id: { type: mongoose.Schema.Types.ObjectId, ref: "Artist", required: true },
    label_id: { type: mongoose.Schema.Types.ObjectId, ref: "Label", default: null },
    album_type: { type: String, enum: ["album", "single", "ep", "compilation"], required: true, default: "album" },
    cover_file_id: { type: mongoose.Schema.Types.ObjectId, ref: "File", default: null },
    release_date: { type: String, default: null },
    total_tracks: { type: Number, required: true, default: 0 },
    total_duration_ms: { type: Number, required: true, default: 0 },
    upc: { type: String, default: null },
    copyright: { type: String, default: null },
    popularity: { type: Number, required: true, default: 0 },
  },
  { timestamps: { createdAt: "created_at", updatedAt: "updated_at" } }
);

albumsSchema.index({ artist_id: 1, release_date: 1 });
albumsSchema.index({ label_id: 1 });
albumsSchema.index({ album_type: 1, release_date: 1 });
albumsSchema.index({ popularity: 1 });
albumsSchema.index({ release_date: 1 });

module.exports = mongoose.model("Album", albumsSchema);

File: schemas/media-streaming/spacetimedb/albums.rs
// albums: Music albums, singles, EPs, and compilations released by artists.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum AlbumType {
    Album, // type: String
    Single,
    Ep,
    Compilation,
}

#[spacetimedb::table(name = albums, public)]
pub struct Album {
    #[primary_key]
    pub id: String, // UUID
    pub title: String,
    #[unique]
    pub slug: String,
    #[index(btree)]
    pub artist_id: String, // UUID — FK → artists.id (cascade delete)
    #[index(btree)]
    pub label_id: Option<String>, // UUID — FK → labels.id (set null on delete)
    pub album_type: AlbumType,
    pub cover_file_id: Option<String>, // UUID — FK → files.id (set null on delete)
    #[index(btree)]
    pub release_date: Option<String>,
    pub total_tracks: i32,
    pub total_duration_ms: i32,
    pub upc: Option<String>,
    pub copyright: Option<String>,
    #[index(btree)]
    pub popularity: i32,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
// Composite index: (artist_id, release_date) — not supported in SpacetimeDB; document only.
// Composite index: (album_type, release_date) — not supported in SpacetimeDB; document only.
